File: TS/系统/09．表现系统/03．垂直滚动条轨道.ts
/**
 * 垂直滚动条轨道（BACKDROP 轨道 + 圆形 thumb）
 * - 仅保留滚轮滑动功能，删除鼠标拖拽
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

const jass = require("jass.common") as any;
const japi = require("jass.japi") as any;

import { getWindowHeight } from "../../lib/扩展函数/封装函数/04．硬件输入/index";
import { setFramePointRelative, FramePoint } from "./01．UI工具/index";

/** thumb 垂直可移动行程（归一化 UI 高度），与 syncThumb 使用同一公式 */
export function getScrollbarThumbTravelNorm(
  trackHeightNorm: number,
  thumbSizeNorm: number,
  topCompensation: number,
  bottomCompensation: number
): number {
  const yRange = trackHeightNorm - thumbSizeNorm + topCompensation + bottomCompensation;
  return Math.max(0, yRange);
}

/** 与 syncThumb 一致的轨道像素高度（Dz 纵向 0..0.6 对应 client 高度） */
export function getScrollbarTrackThumbTravelPx(travelNorm: number): number {
  const ch = (japi as any).DzGetClientHeight();
  const clientH = typeof ch === "number" && ch > 0 ? ch : getWindowHeight() || 600;
  return Math.max(1, (clientH * travelNorm) / 0.6);
}

export interface VerticalScrollbarTrackOptions {
  /** 轨道 BACKDROP */
  trackFrame: number;
  /** thumb BACKDROP */
  thumbFrame: number;
  /** 列表可视区高度（归一化） */
  listViewHeightNorm: number;
  /** 轨道 BACKDROP 实际高度 */
  trackHeightNorm?: number;
  thumbSizeNorm: number;
  topCompensation: number;
  bottomCompensation: number;
  /** 内容总高度（归一化） */
  getTotalContentHeight(): number;
  getScrollOffset(): number;
  setScrollOffset(value: number): void;
  /** 面板是否可交互 */
  isInteractionEnabled(): boolean;
  /** scrollOffset 变化后刷新 */
  onScrollChanged(): void;
  /** 为 true 时不改 thumb 位置 */
  skipManualThumbSync?(): boolean;
}

/** 封装：仅保留 thumb 位置同步，删除鼠标拖拽 */
export class VerticalScrollbarTrack {
  private readonly opt: VerticalScrollbarTrackOptions;

  constructor(options: VerticalScrollbarTrackOptions) {
    this.opt = options;
  }

  /** 初始化（无命中按钮，无拖拽） */
  attach(): void {
    // 仅初始化，不注册鼠标事件
  }

  destroy(): void {
    // 无需清理
  }

  /** 面板隐藏时调用 */
  cancelDrag(): void {
    // 无拖拽状态需要清理
  }

  getHitButtonFrame(): number | null {
    return null;
  }

  isDragging(): boolean {
    return false;
  }

  private trackH(): number {
    const t = this.opt.trackHeightNorm;
    if (t !== undefined && t > 0) return t;
    return this.opt.listViewHeightNorm;
  }

  /** 根据 scrollOffset / maxScroll 设置 thumb 相对轨道位置 */
  syncThumbVisual(maxScroll: number): void {
    (pcall as any)(() => {
      if (typeof jass.GetLocalPlayer !== "function") return;
      const lp = jass.GetLocalPlayer();
      if (lp == null) return;

      if (this.opt.skipManualThumbSync && this.opt.skipManualThumbSync()) return;
      if (!this.opt.thumbFrame || this.opt.thumbFrame === 0) return;
      if (!this.opt.trackFrame || this.opt.trackFrame === 0) return;

      const tf = this.opt.thumbFrame;
      if (typeof (japi as any).DzFrameShow === "function") (japi as any).DzFrameShow(tf, true);
      if (typeof (japi as any).DzFrameSetLevel === "function") (japi as any).DzFrameSetLevel(tf, 120);
      if (typeof (japi as any).DzFrameSetSize === "function") {
        (japi as any).DzFrameSetSize(tf, this.opt.thumbSizeNorm, this.opt.thumbSizeNorm);
      }

      const safeRange = getScrollbarThumbTravelNorm(
        this.trackH(),
        this.opt.thumbSizeNorm,
        this.opt.topCompensation,
        this.opt.bottomCompensation
      );
      const progress = maxScroll <= 0 ? 0 : this.opt.getScrollOffset() / maxScroll;
      const yOffset =
        (0.5 - progress) * safeRange +
        (this.opt.topCompensation - this.opt.bottomCompensation) * 0.5;

      if (typeof (japi as any).DzFrameClearAllPoints === "function") {
        (japi as any).DzFrameClearAllPoints(tf);
      }
      setFramePointRelative(tf, FramePoint.CENTER, this.opt.trackFrame, FramePoint.CENTER, 0, yOffset);
    });
  }
}

File: src/系统/09．表现系统/03．垂直滚动条轨道.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local ____exports = {}
local ____index = require("lib.扩展函数.封装函数.04．硬件输入.index")
local getWindowHeight = ____index.getWindowHeight
local ____index = require("系统.09．表现系统.01．UI工具.index")
local setFramePointRelative = ____index.setFramePointRelative
local FramePoint = ____index.FramePoint
--- 垂直滚动条轨道（BACKDROP 轨道 + 圆形 thumb）
-- - 仅保留滚轮滑动功能，删除鼠标拖拽
local jass = require("jass.common")
local japi = require("jass.japi")
--- thumb 垂直可移动行程（归一化 UI 高度），与 syncThumb 使用同一公式
function ____exports.getScrollbarThumbTravelNorm(self, trackHeightNorm, thumbSizeNorm, topCompensation, bottomCompensation)
    local yRange = trackHeightNorm - thumbSizeNorm + topCompensation + bottomCompensation
    return math.max(0, yRange)
end
--- 与 syncThumb 一致的轨道像素高度（Dz 纵向 0..0.6 对应 client 高度）
function ____exports.getScrollbarTrackThumbTravelPx(self, travelNorm)
    local ch = japi.DzGetClientHeight()
    local clientH = type(ch) == "number" and ch > 0 and ch or (getWindowHeight(nil) or 600)
    return math.max(1, clientH * travelNorm / 0.6)
end
--- 封装：仅保留 thumb 位置同步，删除鼠标拖拽
____exports.VerticalScrollbarTrack = __TS__Class()
local VerticalScrollbarTrack = ____exports.VerticalScrollbarTrack
VerticalScrollbarTrack.name = "VerticalScrollbarTrack"
function VerticalScrollbarTrack.prototype.____constructor(self, options)
    self.opt = options
end
function VerticalScrollbarTrack.prototype.attach(self)
end
function VerticalScrollbarTrack.prototype.destroy(self)
end
function VerticalScrollbarTrack.prototype.cancelDrag(self)
end
function VerticalScrollbarTrack.prototype.getHitButtonFrame(self)
    return nil
end
function VerticalScrollbarTrack.prototype.isDragging(self)
    return false
end
function VerticalScrollbarTrack.prototype.trackH(self)
    local t = self.opt.trackHeightNorm
    if t ~= nil and t > 0 then
        return t
    end
    return self.opt.listViewHeightNorm
end
function VerticalScrollbarTrack.prototype.syncThumbVisual(self, maxScroll)
    pcall(function ()
            if type(jass.GetLocalPlayer) ~= "function" then
                return
            end
            local lp = jass.GetLocalPlayer()
            if lp == nil then
                return
            end
            if self.opt.skipManualThumbSync and self.opt:skipManualThumbSync() then
                return
            end
            if not self.opt.thumbFrame or self.opt.thumbFrame == 0 then
                return
            end
            if not self.opt.trackFrame or self.opt.trackFrame == 0 then
                return
            end
            local tf = self.opt.thumbFrame
            if type(japi.DzFrameShow) == "function" then
                japi.DzFrameShow(tf, true)
            end
            if type(japi.DzFrameSetLevel) == "function" then
                japi.DzFrameSetLevel(tf, 120)
            end
            if type(japi.DzFrameSetSize) == "function" then
                japi.DzFrameSetSize(tf, self.opt.thumbSizeNorm, self.opt.thumbSizeNorm)
            end
            local safeRange = ____exports.getScrollbarThumbTravelNorm(
                nil,
                self:trackH(),
                self.opt.thumbSizeNorm,
                self.opt.topCompensation,
                self.opt.bottomCompensation
            )
            local progress = maxScroll <= 0 and 0 or self.opt:getScrollOffset() / maxScroll
            local yOffset = (0.5 - progress) * safeRange + (self.opt.topCompensation - self.opt.bottomCompensation) * 0.5
            if type(japi.DzFrameClearAllPoints) == "function" then
                japi.DzFrameClearAllPoints(tf)
            end
            setFramePointRelative(
                nil,
                tf,
                FramePoint.CENTER,
                self.opt.trackFrame,
                FramePoint.CENTER,
                0,
                yOffset
            )
        end
    )
end
return ____exports
